src/appleseed/renderer/modeling/object/meshobjectprimitives.cpp

```diff
@@ -0,0 +1,251 @@
+
+// Standard headers.
+#include <cmath>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+class ParametricSphere
+{
+public:
+    explicit ParametricSphere(const double radius)
+      : m_radius(radius)
+    {
+    }
+
+    GVector3 evaluate(const double u, const double v) const
+    {
+        const double theta = 2.0 * Pi * u;
+        const double phi = Pi * v;
+
+        return GVector3(
+            m_radius * cos(theta) * sin(phi),
+           -m_radius * cos(phi),
+            m_radius * sin(theta) * sin(phi)
+        );
+    }
+
+    const double m_radius;
+};
+
+class ParametricGrid
+{
+public:
+    ParametricGrid(const double width, const double height)
+      : m_width(width)
+      , m_height(height)
+    {
+    }
+
+    GVector3 evaluate(const double u, const double v) const
+    {
+        return GVector3(
+            m_width * (u - 0.5),
+            0.0,
+            m_height * (v - 0.5));
+    }
+
+    const double m_width;
+    const double m_height;
+};
+
+class ParametricTorus
+{
+public:
+    ParametricTorus(const double major_radius, const double minor_radius)
+      : m_major_radius(major_radius)
+      , m_minor_radius(minor_radius)
+    {
+    }
+
+    GVector3 evaluate(const double u, const double v) const
+    {
+        const double theta = 2.0 * Pi * u;
+        const double phi = 2.0 * Pi * v;
+
+        return GVector3(
+            (m_major_radius + m_minor_radius * cos(phi)) * cos(theta),
+             m_minor_radius * sin(phi),
+            (m_major_radius + m_minor_radius * cos(phi)) * sin(theta));
+    }
+
+    const double m_major_radius;
+    const double m_minor_radius;
+};
+
+template<typename ParametricSurface>
+void create_vertices(MeshObject& mesh, ParametricSurface surface, const size_t resolution_u, const size_t resolution_v)
+{
+    const size_t num_points = resolution_u * resolution_v;
+
+    mesh.reserve_vertices(num_points);
+    mesh.reserve_vertex_normals(num_points);
+    mesh.reserve_vertex_tangents(num_points);
+    mesh.reserve_tex_coords(num_points);
+
+    // A small step used to calculate the u and v derivatives using forward differencing.
+    const double h = 1.0 / (max(resolution_u, resolution_v) * 4.0);
+
+    for (size_t j = 0; j < resolution_v; ++j)
+    {
+        const double v = fit<size_t, double>(j, 0, resolution_v - 1, 0.0, 1.0);
+        for (size_t i = 0; i < resolution_u; ++i)
+        {
+            const double u = fit<size_t, double>(i, 0, resolution_u - 1, 0.0, 1.0);
+            const GVector3 p = surface.evaluate(u, v);
+
+            // Compute u tangent using forward differencing.
+            GVector3 dpdu = surface.evaluate(u + h, v) - p;
+            const double dpdu_norm = norm(dpdu);
+
+            if (dpdu_norm == 0.0)
+            {
+                // If the u tangent is zero (surface pole), approximate it using a nearby point.
+                dpdu = normalize(surface.evaluate(u + h, v + h) - p);
```

Comment:
I added validation on the parameters and I think it cannot happen anymore with the primitives implemented.